### PR TITLE
client-api-react with subscriptions to selections and labels; onSelection

### PR DESCRIPTION
**Pull request by exrhizo**
Updates for a subscribe mechanism, for selection and labels, examples, storybooks and docs for these.

requires: https://github.com/graphistry/graphistry/pull/1764

# Running locally
```
const LOCAL_DEV = {
  graphistryHost: "http://0.0.0.0:3000",
  play: 0,
  session: "cycle"
}
```
and with`<Graphistry {...LOCAL_DEV} />` in the App.js of Create React App test

### Style changes to demo app

Started with getting the `cra-test` demo app style cleaned up.

image

I changed host to be a local instance of graphistry viz, while I figure out how to expose the selection

### CODE REVIEW

File: projects/client-api-react/src/index.js [hunk 1]
```diff
@@ -479,6 +490,24 @@ function Graphistry(props) {
...
+            onUpdateObservableG({observable: gObs, subscription: gSub});
+            return () => onUpdateObservableG(null);
+        }
+    }, [gObs, onUpdateObservableG]);
+
+    useEffect(() => {
+        if (g && onSelectionUpdate) {
+            const sub = selectionUpdates(g).subscribe(
+                (v) => onSelectionUpdate(v),
+                (error) => console.error('client-api-react selectionUpdates subscription error', error)
```
Comment (lmeyerov) [inline projects/client-api-react/src/index.js:504]: maybe we do node style here too, `onSelectionUpdate(err, val)` ?

File: projects/client-api/src/index.js [hunk 2]
```diff
@@ -1303,6 +1303,31 @@ export function updateZoom(level) {
 }
 chainList.updateZoom = updateZoom;
 
+/**
+ * Subscribe to label change and exit events
+ * @function selectionUpdates
+ * @param {object} [cache] - An optional cache object (empty object) to use for caching label updates across subscriptions.
+ * @return {Subscription} A {@link Subscription} that can be used to stop reacting to label updates
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1310]: can you more explicitly clarify what gets received in docstr + example?

**Review by lmeyerov — COMMENTED** [review]
afaict mostly just docstr tweak for this one

making subscription opt-in on streamgl-viz side may require a bit more here

follow-on pr can work through unsubscribe

File: projects/client-api-react/src/stories/Graphistry.stories.jsx [hunk 3]
```diff
@@ -13,9 +13,11 @@ export const Empty = (args) => <Graphistry {...args} />;
 export const PredefinedDataset = (args) => <Graphistry {...args} dataset='Miserables' showSplashScreen={true} />;
 
 const defaultSettings = {
+  // graphistryHost: "http://0.0.0.0:3000",
   dataset: 'Miserables',
   play: 1,
-  showSplashScreen: true
+  showSplashScreen: true,
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/Graphistry.stories.jsx:19]: do the `onXUpdate=undefined` here so works for all?

File: projects/client-api-react/src/stories/Graphistry.stories.jsx [hunk 4]
```diff
@@ -33,6 +35,31 @@ export const OnClientAPIConnected = (args) => {
   </div>);
 }
 
+export const OnSelectionUpdate = (args) => {
+  const [ selection, setSelection ] = useState(undefined);
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/Graphistry.stories.jsx:39]: good to show what the sample return is (nodes, edges, their values) so dev gets to see the schema

File: projects/client-api-react/src/stories/Graphistry.stories.jsx [hunk 5]
```diff
@@ -33,6 +35,31 @@ export const OnClientAPIConnected = (args) => {
...
+  
+  return (<div>
+    { `Selection: ${JSON.stringify(selection)}`}
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/Graphistry.stories.jsx:42]: maybe also console.log?

File: projects/client-api-react/src/stories/Graphistry.stories.jsx [hunk 6]
```diff
@@ -33,6 +35,31 @@ export const OnClientAPIConnected = (args) => {
...
+      onLabelsUpdate={undefined}
+    />
+  </div>);
+}
+
+export const OnLabelUpdate = (args) => {
+  const [ labels, setLabels ] = useState(undefined);
+  
+  return (<div>
+    { `Labels: ${JSON.stringify(labels)}`}
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/Graphistry.stories.jsx:55]: and console.log?

File: projects/client-api-react/src/stories/GraphistryJS.stories.jsx [hunk 7]
```diff
@@ -21,9 +21,8 @@ import {
     tap,
     delay
 } from '@graphistry/client-api';
-import { addExclusion } from '@graphistry/client-api/src';
 
-const basePath = 'https://hub.graphistry.com';
+const basePath = 'http://0.0.0.0:3000';
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/GraphistryJS.stories.jsx:25]: ?

File: projects/client-api-react/src/stories/GraphistryJS.stories.jsx [hunk 8]
```diff
@@ -88,6 +87,50 @@ export const InstantiateGraphistryJS = (args) => {
     );
 }
 
+// export const GraphistryJSSubscribeToSelection = (args) => {
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/GraphistryJS.stories.jsx:90]: TODO

File: projects/client-api/src/index.js [hunk 9]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+ * @example
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1314]: TODO

File: projects/client-api/src/index.js [hunk 10]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+ *     .pipe(
+ *          selectionUpdates,
+ *          tap((s) => {
+ *                 console.log(s);
+ *          }),
+ *     })
+ *     .subscribe();
+ */
+export function selectionUpdates(g) {
+    console.log('client-api.APISUB', 'selectionUpdates', g, 'exor');
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1325]: DEBUG

File: projects/client-api/src/index.js [hunk 11]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+ * GraphistryJS(document.getElementById('viz'))
+ *     .pipe(
+ *          selectionUpdates,
+ *          tap((s) => {
+ *                 console.log(s);
+ *          }),
+ *     })
+ *     .subscribe();
+ */
+export function selectionUpdates(g) {
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1324]: should also export elsewhere similar to labels

File: projects/client-api/src/index.js [hunk 12]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+    return g.selectionStream || (g.selectionStream = new BehaviorSubject('value'))
+        .pipe(
+            tap((v) => {
+                console.debug('client-api.APISUB', 'post graphistry-subscribe', 'exor');
+                g.iFrame.contentWindow.postMessage({ type: 'graphistry-subscribe', agent: 'graphistryjs', path: SELECTION_PATH }, '*');
+            }),
+            finalize(() => {
+                console.debug('client-api.APISUB', 'finalize post graphistry-unsubscribe', 'exor');
+                g.iFrame.contentWindow.postMessage({ type: 'graphistry-unsubscribe', agent: 'graphistryjs', path: SELECTION_PATH }, '*');
+            }),
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1340]: i'm unsure of relationship with shareReplay: should test

ex: if unsub (ex: close a panel), does it actually unsub (no child pushes?)
ex: if resub on a closed, does it actually resub?

File: projects/client-api/src/index.js [hunk 13]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+            switchMap(() =>
+                fromEvent(window, 'message').pipe(
+                    map(o => o.data),
+                    tap(o => console.log('client-api.APISUB', 'recieved message', o, 'exor')),
+                    filter(o => o && o.type === 'graphistry-update' && o.path === SELECTION_PATH),
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1345]: rename `graphistry-sub-update` b/c these are only from the sub channel?

File: projects/client-api/src/index.js [hunk 14]
```diff
@@ -1303,6 +1306,71 @@ export function updateZoom(level) {
...
+                    map(o => o.data),
+                    tap(({ edge, point}) => {
+                        g.models.model.setCache({
```
Comment (lmeyerov) [inline projects/client-api/src/index.js:1348]: don't need a subscribe?

**Review by lmeyerov — CHANGES_REQUESTED** [review]
see inline

**Comment by lmeyerov** [comment]
also doc strs

File: projects/client-api-react/src/index.js (hunk 1 above)
Comment (exrhizo) [reply projects/client-api-react/src/index.js]: done.

File: projects/client-api-react/src/stories/Graphistry.stories.jsx (hunk 3 above)
Comment (exrhizo) [reply projects/client-api-react/src/stories/Graphistry.stories.jsx:19]: yes.

File: projects/client-api/src/index.js (hunk 12 above)
Comment (exrhizo) [reply projects/client-api/src/index.js:1340]: Sorted this out, and it works in the storybook example

File: projects/client-api/src/index.js (hunk 14 above)
Comment (exrhizo) [reply projects/client-api/src/index.js:1348]: no, I checked it works to do a read.

Also, setCache returns the model, not an observable.

File: projects/client-api-react/src/stories/Graphistry.stories.jsx [hunk 15]
```diff
@@ -5,19 +5,23 @@ import { Graphistry } from '../index';
...
 export const Empty = (args) => <Graphistry {...args} />;
 export const PredefinedDataset = (args) => <Graphistry {...args} dataset='Miserables' showSplashScreen={true} />;
 
 const defaultSettings = {
-  // graphistryHost: "http://0.0.0.0:3000",
+  graphistryHost: "http://0.0.0.0:3000", // TODO: remove
+  session: "cycle",
   dataset: 'Miserables',
-  play: 1,
+  play: 0,
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/Graphistry.stories.jsx:23]: keep `1` ?

Comment (exrhizo) [reply projects/client-api-react/src/stories/Graphistry.stories.jsx:23]: reverted

File: projects/client-api-react/src/stories/Graphistry.stories.jsx (hunk 6 above)
Comment (exrhizo) [reply projects/client-api-react/src/stories/Graphistry.stories.jsx:58]: skipping

File: projects/client-api/src/index.js (hunk 9 above)
Comment (exrhizo) [reply projects/client-api/src/index.js:1314]: fixed

File: projects/client-api/src/index.js (hunk 11 above)
Comment (exrhizo) [reply projects/client-api/src/index.js:1322]: done.

File: projects/client-api/src/index.js (hunk 13 above)
Comment (exrhizo) [reply projects/client-api/src/index.js:1343]: done

File: projects/client-api-react/src/stories/GraphistryJS.stories.jsx [hunk 16]
```diff
@@ -88,6 +95,108 @@ export const InstantiateGraphistryJS = (args) => {
     );
 }
 
+const SUBSCRIPTION_TEST = [
```
Comment (lmeyerov) [inline projects/client-api-react/src/stories/GraphistryJS.stories.jsx:98]: this test is useful but maybe not for standard storybook docs... by default don't export, and we only export in dev via manual code change? or make a separate stories file for testing?

**Comment by lmeyerov** [comment]
Also to check: are js + react docs looking ok?

* js: https://hub.graphistry.com/static/js-docs/jsdocs/index.html
* react: https://hub.graphistry.com/static/js-docs/index.html?path=/docs/graphistry-react-style--empty

**Review by exrhizo — COMMENTED** [review]
Okay, this feels ready, after one more go over.

**Pull request merged**